internal/schema/schema.go

```diff
@@ -0,0 +1,332 @@
+package schema
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/qri-io/jsonschema"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	enginev1 "github.com/cerbos/cerbos/api/genpb/cerbos/engine/v1"
+	schemav1 "github.com/cerbos/cerbos/api/genpb/cerbos/schema/v1"
+	"github.com/cerbos/cerbos/internal/config"
+	"github.com/cerbos/cerbos/internal/observability/logging"
+	"github.com/cerbos/cerbos/internal/storage"
+)
+
+type Manager struct {
+	conf                *Conf
+	log                 *zap.SugaredLogger
+	store               storage.Store
+	principalProperties map[string]string
+	resourceProperties  map[string]map[string]string
+	schema              *schemav1.Schema
+	principalSchema     *jsonschema.Schema
+	resourceSchemas     map[string]*jsonschema.Schema
+	mu                  sync.RWMutex // mutex for schema, principalSchema and resourceSchemas fields
+}
+
+func New(store storage.Store) (*Manager, error) {
+	conf := &Conf{}
+	if err := config.GetSection(conf); err != nil {
+		return nil, fmt.Errorf("failed to get config section %q: %w", confKey, err)
+	}
+
+	mgr := &Manager{
+		conf:                conf,
+		log:                 zap.S().Named("schema"),
+		store:               store,
+		principalProperties: make(map[string]string),
+		resourceProperties:  make(map[string]map[string]string),
+	}
+
+	err := mgr.ReadOrUpdateSchemaFromStore()
+	if err != nil {
+		var msg = fmt.Sprintf("failed to read schema file from store: %s", err)
+		if conf.Enforcement == EnforcementReject {
+			return nil, fmt.Errorf(msg)
+		} else {
+			mgr.log.Warn(msg)
+		}
+	}
+
+	store.Subscribe(mgr)
+
+	return mgr, nil
+}
+
+func (m *Manager) Validate(ctx context.Context, input *enginev1.CheckInput) error {
+	if m.schema == nil {
```

Comment:
Whenever you access `schema`, `principalSchema` or `resourceSchema` it has to be protected by a read lock ( `mu.RLock`) otherwise the mutex is pointless. I am only commenting here but the issue persists across all of the methods below.